# 01. Importing Libraries

In [1]:
# Import the pandas, NumPy, and os libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Creating a path to the root directory for this project
path = r'C:\Users\jpelc\Documents\01-2023 Instacart Basket Analysis'

In [3]:
# Importing orders.csv 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = [0])

In [4]:
# Importing products.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 03. Data Consistency

### Mixed-Type Data

In [5]:
# Mixed-Type Data
# Create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Fixing mixed types
df_test['mix'] = df_test['mix'].astype('str')

### Missing Values

In [10]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# View missing values in 'product_name'
df_nan = df_prods[df_prods['product_name'].isnull() ==True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Use df.describe() to find the mean
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

### Replacing missing values

In [13]:
# To repalce missing values use the following command
# df['column with missings'].fillna(mean value, inplace=True)

In [14]:
# Find the median
# df_prods.median()

In [15]:
# Calculate size of the dataframe
df_prods.shape

(49693, 5)

In [16]:
# Creating a clean database
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Calculate size of the dataframe
df_prods_clean.shape

(49677, 5)

### Duplicates

In [18]:
# Finding duplicates in df_prods
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
# Calculate size of the dataframe
df_prods_clean.shape

(49677, 5)

In [21]:
# Addressing Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [23]:
#1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The max value for 'prices' is 99999. Needs to be checked for errors.

In [24]:
# Looking for any product with value of $100 or more
df_prods_clean_no_dups[df_prods_clean_no_dups["prices"]>= 100]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

The price seems to be off by 4 decimal points by my best guess. So the value of cottage cheese should be  1.49, and Reduced fat milk at  9.99.  But I will refrain on making any changes, and follow up with someone that can answer this question first.

In [25]:
#2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, 
#   share in a markdown cell whether anything about the data looks off or should be investigated further. 
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Nothing very odd about the dataframe. Orders are made during 6 days, sundays I am assuming they are closed. The amount of orders seems to be from 1 to 100 and nothing over that. They seem to take orders 24hrs a day, and most days since prior order is 30 days(one month).  

In [26]:
#3. Check for mixed-type data in your df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed type data columns where found.

In [27]:
#5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the 'days_since_prior_order' column. 

In [28]:
# creating a dataframe to view the missing values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

Looks like the first order for every user does not contain a value in 'days_since_last_order'. The max'user_id' coincides with the amount of dublicates found.

In [29]:
#6. Address the missing values using an appropriate method.

I will be leaving the dataframe as is. The missing values are the first order of each user, statistics can be calculated with those missing value, and the other information in these rows is valuable. Imputing values would be an option due to the missing values only amounting to roughly 6% of the data.

In [30]:
#7. Run a check for duplicate values in your df_ords data.
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates where found.

In [31]:
#9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [32]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))